idx/memory/tag_query.go

```diff
@@ -0,0 +1,411 @@
+	}
+
+	key := expr[:pos]
+
+	// if !
+	if expr[pos] == 33 {
+		not = true
+		pos++
+	}
+
+	// expecting a =
+	if expr[pos] != 61 {
+		return expression{operator: PARSING_ERROR}
+	}
+	pos++
+
+	// if ~
+	if len(expr) > pos && expr[pos] == 126 {
+		regex = true
+		pos++
+	}
+
+	valuePos := pos
+	for ; pos < len(expr); pos++ {
+		// disallow ; in value
+		if expr[pos] == 59 {
+			return expression{operator: PARSING_ERROR}
+		}
+	}
+	value := expr[valuePos:]
+
+	if not {
+		if regex {
+			return expression{kv: kv{key: string(key), value: string(value)}, operator: NOT_MATCH}
+		} else {
+			return expression{kv: kv{key: string(key), value: string(value)}, operator: NOT_EQUAL}
+		}
+	} else {
+		if regex {
+			return expression{kv: kv{key: string(key), value: string(value)}, operator: MATCH}
+		} else {
+			return expression{kv: kv{key: string(key), value: string(value)}, operator: EQUAL}
+		}
+	}
+}
+
+func NewTagQuery(expressions []string, from int64) (TagQuery, error) {
+	q := TagQuery{from: from}
+
+	if len(expressions) == 0 {
+		return q, errInvalidQuery
+	}
+
+	for _, expr := range expressions {
+		e := parseExpression(expr)
+		if e.operator == PARSING_ERROR {
+			return q, errInvalidQuery
+		}
+
+		// special case of empty value
+		if len(e.value) == 0 {
+			if e.operator == EQUAL || e.operator == MATCH {
+				q.notMatch = append(q.notMatch, kv{
+					key:   e.key,
+					value: "^.+",
+				})
+			} else {
+				q.match = append(q.match, kv{
+					key:   e.key,
+					value: "^.+",
+				})
+			}
+		} else {
+			// always anchor all regular expressions at the beginning
+			if (e.operator == MATCH || e.operator == NOT_MATCH) && len(e.value) > 0 && e.value[0] != byte('^') {
+				e.value = "^" + e.value
+			}
+
+			switch e.operator {
+			case EQUAL:
+				q.equal = append(q.equal, e.kv)
+			case NOT_EQUAL:
+				q.notEqual = append(q.notEqual, e.kv)
+			case MATCH:
+				q.match = append(q.match, e.kv)
+			case NOT_MATCH:
+				q.notMatch = append(q.notMatch, e.kv)
+			}
+		}
+	}
+
+	return q, nil
+}
+
+func (q *TagQuery) getInitialByMatch(index TagIndex) (int, TagIDs, error) {
+	resultSet := make(TagIDs)
+	lowestCount := math.MaxUint32
+	startId := 0
+
+	// choose key that has the smallest number of values because we want to
+	// reduce the number of matches and start with the smallest possible resultSet.
+	// the smaller the resultSet the less time will have to be spent to further
+	// filter its values in the following expressions.
+	for i := range q.match {
+		l := len(index[q.match[i].key])
+		if l < lowestCount {
+			lowestCount = l
+			startId = i
+		}
+	}
+
+	// shortcut if pattern is ^.+ (f.e. expression "key!=~" will be translated to "key=~^.+")
+	if q.match[startId].value == "^.+" {
+		for _, ids := range index[q.match[startId].key] {
+			for id := range ids {
+				resultSet[id] = struct{}{}
+			}
+		}
+		return startId, resultSet, nil
+	}
+
+	re, err := regexp.Compile(q.match[startId].value)
+	if err != nil {
+		return 0, nil, errInvalidQuery
```

Comment:
similar as above, can we validate the query before trying to run it (and holding a lock) ?

Reply:
done: https://github.com/grafana/metrictank/pull/729/commits/ecaa3fb114e96e6c29b731db2239831460a83486
performance is comparable:
```
before:
BenchmarkTagQueryFilterAndIntersect-8            	     200	  71858744 ns/op	10348793 B/op	   95083 allocs/op
BenchmarkTagQueryFilterAndIntersectOnlyRegex-8   	      50	 377291417 ns/op	38070374 B/op	  971222 allocs/op

after:
BenchmarkTagQueryFilterAndIntersect-8            	     200	  67579220 ns/op	10349799 B/op	   95086 allocs/op
BenchmarkTagQueryFilterAndIntersectOnlyRegex-8   	      50	 385191954 ns/op	38067751 B/op	  971222 allocs/op
```

Reply:
i would only expect a performance difference when we would have hit invalid queries under lock contention (concurrency) but anyway the change is good. cleaner too.